storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java

```diff
@@ -29,25 +35,768 @@
+                        throw new OffsetOutOfRangeException("Received request for offset " + startOffset + " for partition " + topicPartition + ", " +
+                                "but we only have log segments upto " + endOffset + ".");
+                    }
+                    if (startOffset == maxOffsetMetadata.messageOffset) {
+                        return emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns);
+                    } else if (startOffset > maxOffsetMetadata.messageOffset) {
+                        return emptyFetchDataInfo(convertToOffsetMetadataOrThrow(startOffset), includeAbortedTxns);
+                    } else {
+                        // Do the read on the segment with a base offset less than the target offset
+                        // but if that segment doesn't contain any messages with an offset greater than that
+                        // continue to read from successive segments until we get some messages or we reach the end of the log
+                        FetchDataInfo fetchDataInfo = null;
+                        while (fetchDataInfo == null && segmentOpt.isPresent()) {
+                            LogSegment segment = segmentOpt.get();
+                            long baseOffset = segment.baseOffset();
+
+                            // 1. If `maxOffsetMetadata#segmentBaseOffset < segment#baseOffset`, then return maxPosition as empty.
+                            // 2. Use the max-offset position if it is on this segment; otherwise, the segment size is the limit.
+                            // 3. When maxOffsetMetadata is message-offset-only, then we don't know the relativePositionInSegment so
+                            //    return maxPosition as empty to avoid reading beyond the max-offset
+                            Optional<Long> maxPositionOpt;
+                            if (segment.baseOffset() < maxOffsetMetadata.segmentBaseOffset)
+                                maxPositionOpt = Optional.of((long) segment.size());
+                            else if (segment.baseOffset() == maxOffsetMetadata.segmentBaseOffset && !maxOffsetMetadata.messageOffsetOnly())
+                                maxPositionOpt = Optional.of((long) maxOffsetMetadata.relativePositionInSegment);
+                            else
+                                maxPositionOpt = Optional.empty();
+
+                            fetchDataInfo = segment.read(startOffset, maxLength, maxPositionOpt, minOneMessage);
+                            if (fetchDataInfo != null) {
+                                if (includeAbortedTxns) {
+                                    fetchDataInfo = addAbortedTransactions(startOffset, segment, fetchDataInfo);
+                                }
+                            } else {
+                                segmentOpt = segments.higherSegment(baseOffset);
+                            }
+                        }
+
+                        if (fetchDataInfo != null) {
```

Comment:
IntelliJ was suggesting me that change too. Maybe it's because I'm not used to see this helper in code but I find the simple `if/else` block easier to read.